### PR TITLE
SONARKT-291 Implement rule S6516: Functional interface implementations should use lambda expressions

### PR DESCRIPTION
**Pull request by kaufco**

### CODE REVIEW

**Review by leveretka — CHANGES_REQUESTED** [review]

File: sonar-kotlin-plugin/src/test/java/org/sonarsource/kotlin/checks/SamConversionCheckTest.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        val mockContext = mockk<KotlinFileContext>()
+
+        assertDoesNotThrow {
+            val check = SamConversionCheck()
+            check.visitObjectDeclaration(mockDeclaration, mockContext)
+        }
+    }
+
+    @Test
+    fun testFlexibleType() {
```
[the remark this answers is not in the document]
Comment (kaufco) [reply sonar-kotlin-plugin/src/test/java/org/sonarsource/kotlin/checks/SamConversionCheckTest.kt:54]: Fixed.

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SamConversionCheck.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.sonarsource.kotlin.api.isFunctionalInterface
+import org.sonarsource.kotlin.api.merge
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+@Rule(key = "S6516")
+class SamConversionCheck : AbstractCheck() {
+
+    override fun visitObjectDeclaration(declaration: KtObjectDeclaration, context: KotlinFileContext) {
+        val superTypeEntry = declaration.superTypeListEntries.singleOrNull() ?: return
+        val superType = superTypeEntry.typeReference?.getType(context.bindingContext) as? SimpleType ?: return
```
[the remark this answers is not in the document]
Comment (kaufco) [reply sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SamConversionCheck.kt:37]: Fixed.

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt [hunk 3]
```diff
@@ -492,9 +496,28 @@ fun KtExpression?.getterMatches(bindingContext: BindingContext, propertyName: St
...
-fun PsiElement?.getType(bindingContext: BindingContext) =
+fun PsiElement?.getVariableType(bindingContext: BindingContext) =
     this?.let { bindingContext.get(BindingContext.VARIABLE, it)?.type }
 
+fun KtTypeReference?.getType(bindingContext: BindingContext): KotlinType? =
+    this?.let { bindingContext.get(BindingContext.TYPE, it) }
+
+fun KtClassOrObject.hasExactlyOneFunctionAndNoProperties(): Boolean {
+    // Note: other possible declarations are KtClass (classes, interfaces) and
+    //       KtObjectDeclaration (companion objects), but they are allowed inside function interfaces
```
[the remark this answers is not in the document]
Comment (kaufco) [reply sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt:507]: Fixed.

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt [hunk 4]
```diff
@@ -492,9 +496,28 @@ fun KtExpression?.getterMatches(bindingContext: BindingContext, propertyName: St
...
+    var functionCount = 0
+    return declarations.all {
+        it !is KtProperty && (it !is KtNamedFunction || functionCount++ == 0)
+    } && functionCount > 0
+}
+
+fun KotlinType.isFunctionalInterface(): Boolean {
+    return getSingleAbstractMethodOrNull(constructor.declarationDescriptor as ClassDescriptor) != null
```
[the remark this answers is not in the document]
Comment (kaufco) [reply sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt]: The function itself is general enough for the APIExtensions I think. A service function that returns whether a KotlinType is a functional interface might come in handy. 

The question is whether the implementation is general enough, too. I'm using `getAbstractSingleMethodOrNull` here, which comes from `org.jetbrains.kotlin.resolve.sam`, and it's assumption is that we have a `ClassDescriptor` available.

The case that `declarationConstructor` is either `null` or not a `ClassDescriptor` is now handled correctly and `false` is returned in that case. What still bothers me is: When can it be null? Probably when the type does not belong to the compilation source but comes from a JAR? Wdyt?

**Review by leveretka — APPROVED** [review]
Nice work on the rule, looks good!

**Pull request merged**